### ISSUE TITLE
Using extreme_deconvolution library in XDGMM?

### ISSUE THREAD

**Issue by garmilla** [astroML/astroML#50]
Hello. I've been using the XDGMM code to estimate density distributions in color-color diagrams. It worked well except that it took very long to converge. At some point I heard that Bovy et al. put a link in their paper to the code they wrote. This is the link

https://code.google.com/p/extreme-deconvolution/

I installed their extreme_deconvolution library and their python wrapper and did this 

```python
import extreme_deconvolution as xd
from sklearn.mixture import GMM
from astroML.density_estimation import XDGMM

# Load data and measurement covariates in XTrain and XerrTrain
# Choose number of gaussians `n_components`

# Stole this from XDGMM
gmm = GMM(n_components, n_iter=10, covariance_type='full').fit(XTrain)
amp = gmm.weights_; mean = gmm.means_; covar = gmm.covars_

# Results are saved in `amp`, `mean`, and `covar`
xd.extreme_deconvolution(XTrain,XerrTrain,amp,mean,covar)

# Load results in XDGMM object
clf = XDGMM(n_components)
clf.alpha = amp; clf.mu = mean; clf.V = covar

```

I found this to be much much faster than the current code in XDGMM, so I wanted to share this here. I thought it may be a good idea to use their library in XDGMM (The library also has the projection matrix feature which is another good thing). I guess that shipping astroML with another library may create more dependencies and other potential complications. In any case, I hope this is useful and let me know if I can be of any help.

**Comment by jakevdp** [comment]
Yes, the current XDGMM code was meant as a place-holder; one of the desired features of astroML was lightweight installation; i.e. no requirement for compilers on your system. It would definitely be nice to have a faster version, whether that means bundling Bovy's code or creating a new cython implementation. I hope to have the time to work on that some day!

**Comment by garmilla** [comment]
I didn't know astroML was trying to avoid compilers. Probably cython is the best way to speed it up without a compiler. I had a feeling that this was unlikely to have been overlooked. Thanks!

**Comment by jakevdp** [comment]
We're trying to avoid compilers in the base package; there's also ``astroML_addons`` where cython and C code could go.

**Comment by jakevdp** [comment]
Also, the XD package is GPL, so we can't use it in astroML :disappointed:

**Comment by garmilla** [comment]
I think the license changed. The code is now on github and the license I see there is not GPL

https://github.com/jobovy/extreme-deconvolution

**Comment by garmilla** [comment]
Although XD uses GSL, so I guess it can't be used.

**Comment by jakevdp** [comment]
I hadn't seen that! In that case, this could be a great addition to ``astroML_addons``: optionally use the fast compiled code if available. Or maybe better, have it call the xd library if it's installed.
